File: apps/web/src/api/couple.ts
import { apiClient } from './client'

/**
 * Paartherapie (peer-to-peer): zwei Nutzer:innen verbinden sich zu einem Paarraum.
 *
 * Wichtig: Eine Kopplung ist KEINE Freigabe. Sie öffnet ausschließlich den gemeinsamen
 * Paarraum und gewährt keinerlei Zugriff auf Fall-Inhalte der anderen Person. Was Echo
 * im Paarraum weiß, stellt der Nutzer später ausdrücklich selbst zusammen.
 */
export interface CoupleLink {
  id: string
  status: 'pending' | 'active' | 'ended'
  role: 'initiator' | 'partner'
  /** Nur sichtbar, solange die eigene Einladung offen ist. */
  invite_code: string | null
  /** Eigener Anker-Fall (nur Bezug, kein Datenzugriff der anderen Person). */
  case_id: string | null
  partner_display_name: string | null
  partner_connected: boolean
  created_at: string
  accepted_at: string | null
}

export interface CoupleInvitePublic {
  valid: boolean
  status: string
}

export interface CoupleAcceptResult {
  connected: boolean
  already: boolean
  couple_id: string | null
}

/**
 * Fortschritt des Paarraums. Bewusst ohne Rangliste: eigene Punkte, gemeinsame Punkte,
 * Meilensteine – aber kein Sieger zwischen euch beiden.
 */
export interface CoupleProgress {
  total_points: number
  own_points: number
  members: { user_id: string; name: string; points: number }[]
  streak_weeks: number
  level: { name: string; next_at: number | null; next_name: string | null }
  milestones: { key: string; title: string; description: string; reached: boolean }[]
  recent: { kind: string; label: string; points: number; name: string; created_at: string }[]
}

export const coupleApi = {
  progress: (coupleId: string) =>
    apiClient.get<CoupleProgress>(`/couple/links/${coupleId}/progress`).then(r => r.data),

  list: () => apiClient.get<CoupleLink[]>('/couple/links').then(r => r.data),
  get: (coupleId: string) =>
    apiClient.get<CoupleLink>(`/couple/links/${coupleId}`).then(r => r.data),
  create: (caseId?: string | null) =>
    apiClient.post<CoupleLink>('/couple/links', { case_id: caseId ?? null }).then(r => r.data),
  accept: (code: string, caseId?: string | null) =>
    apiClient
      .post<CoupleAcceptResult>('/couple/links/accept', { code, case_id: caseId ?? null })
      .then(r => r.data),
  /** Ohne purge nur schließen; mit purge werden die Inhalte wirklich gelöscht. */
  end: (coupleId: string, purge = false) =>
    apiClient.delete(`/couple/links/${coupleId}`, { params: { purge } }),

  /** Löscht nur, was allein dir gehört – Geteiltes bleibt. */
  deleteMyPrivateContent: (coupleId: string) =>
    apiClient.delete(`/couple/links/${coupleId}/my-private-content`),
  checkCode: (code: string) =>
    apiClient
      .get<CoupleInvitePublic>(`/couple/invites/${encodeURIComponent(code)}`)
      .then(r => r.data),
}

/** 8-stelligen Kopplungscode zur Anzeige gruppieren (XXXX-XXXX). */
export function formatCoupleCode(code: string): string {
  return code.length === 8 ? `${code.slice(0, 4)}-${code.slice(4)}` : code
}

/** Einladungslink zum Teilen (Code steckt im Pfad). */
export function coupleInviteLink(code: string): string {
  return `${window.location.origin}/paar/beitreten/${code}`
}

File: services/api/app/schemas/couple.py
"""Schemas: Paartherapie (peer-to-peer) — Kopplung zweier Nutzer:innen.

Eine Person erzeugt eine Einladung (Kopplungscode), die andere nimmt sie an; beide
teilen sich danach einen Paarraum. Eine Kopplung ist KEINE Freigabe: sie gewährt
keinerlei Zugriff auf Fall-Inhalte der jeweils anderen Person. Entsprechend enthalten
diese Schemas bewusst keine Fall-Daten — nur Kopplungs-Metadaten.
Siehe services/couple_therapy_service.py.
"""
from __future__ import annotations

from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field


class CoupleLinkCreate(BaseModel):
    """Einladung erzeugen. Der Anker-Fall ist nur Herkunft/Bezug — kein Datenzugriff."""
    case_id: UUID | None = None


class CoupleLinkCaseUpdate(BaseModel):
    """Welcher eigene Fall zu diesem Paarraum gehoert — oder keiner mehr (None).

    Reine Herkunftsangabe, KEIN Datenzugriff: Sie entscheidet nur, wohin eine im
    Paarraum entstandene Szene gespeichert werden darf.
    """

    case_id: UUID | None = None


class CoupleLinkAccept(BaseModel):
    """Einladung per Kopplungscode annehmen (optional mit eigenem Anker-Fall)."""
    code: str = Field(..., min_length=4, max_length=32)
    case_id: UUID | None = None


class CoupleLinkResponse(BaseModel):
    """Sicht einer Person auf ihren Paarraum bzw. ihre Einladung.

    ``invite_code`` wird nur ausgeliefert, solange die Einladung offen ist.
    ``partner_display_name`` ist der selbstgewählte Anzeigename — keine Fall-Daten.
    """
    id: UUID
    status: str
    role: str                                  # 'initiator' | 'partner'
    invite_code: str | None = None
    case_id: UUID | None = None                # eigener Anker-Fall
    partner_display_name: str | None = None
    partner_avatar: str | None = None
    partner_connected: bool = False
    created_at: datetime
    accepted_at: datetime | None = None


class CoupleLinkAcceptResponse(BaseModel):
    connected: bool
    already: bool = False
    couple_id: UUID | None = None


class CoupleMemberPoints(BaseModel):
    user_id: UUID
    name: str
    points: int


class CoupleMilestone(BaseModel):
    key: str
    title: str
    description: str
    reached: bool


class CoupleProgressEvent(BaseModel):
    kind: str
    label: str
    points: int
    name: str
    created_at: datetime


class CoupleLevel(BaseModel):
    name: str
    next_at: int | None = None
    next_name: str | None = None


class CoupleProgress(BaseModel):
    """Fortschritt eines Paarraums — eigene UND gemeinsame Punkte, bewusst ohne Rangliste."""
    total_points: int
    own_points: int
    members: list[CoupleMemberPoints]
    streak_weeks: int
    level: CoupleLevel
    milestones: list[CoupleMilestone]
    recent: list[CoupleProgressEvent]


class CoupleDashboardItem(BaseModel):
    """Ein Eintrag im Dashboard — entweder wartet er auf dich oder auf die andere Person."""
    kind: str
    title: str
    detail: str
    target: str | None = None


class CoupleDashboardSession(BaseModel):
    id: UUID
    title: str
    status: str
    scheduled_for: datetime | None = None
    message_count: int
    has_summary: bool
    from_topic: bool = False


class CoupleDashboardTopic(BaseModel):
    id: UUID
    title: str
    status: str
    message_count: int
    has_mediation: bool
    open_bridges: int


class CoupleAgreementSnippet(BaseModel):
    id: UUID
    body: str
    status: str


class CoupleAgreementSummary(BaseModel):
    proposed: int
    active: int
    kept: int
    recent: list[CoupleAgreementSnippet] = []


class CoupleEchoSummarySnippet(BaseModel):
    """Eigene Begleiter-Zusammenfassung für die Übersicht."""
    id: UUID
    title: str | None = None
    summary_text: str
    created_at: datetime


class CoupleDashboard(BaseModel):
    """Was im Paarraum gerade dran ist — serverseitig sortiert nach „wer ist am Zug“."""
    partner_name: str | None = None
    partner_avatar: str | None = None
    own_name: str = "Du"
    own_avatar: str | None = None
    echo_summaries: list[CoupleEchoSummarySnippet] = []
    attention: list[CoupleDashboardItem] = []
    waiting_for_partner: list[CoupleDashboardItem] = []
    sessions: list[CoupleDashboardSession] = []
    topics: list[CoupleDashboardTopic] = []
    agreements: CoupleAgreementSummary
    progress: CoupleProgress


class CoupleInvitePublic(BaseModel):
    """Minimal-Sicht auf einen Kopplungscode — bewusst ohne Namen oder Inhalte."""
    valid: bool
    status: str
